### DISCUSSION TITLE
What does this mean? Error: No Makefile to build dependency /home/me/github/rabbitmq/deps/recon ?

### DISCUSSION THREAD

**Discussion by chrispynz4** [rabbitmq/rabbitmq-server#9520] [Questions]
Hi,

I'm trying to make a custom plugin but the plugin needs to be for an old version of RabbitMQ (3.8 or 3.9) since that's what we have in production.

I haven't had much luck building rabbitmq-message-timestamp or rabbitmq-routing-node-stamp so I thought I'd just build everything and try to adapt from one of the built-in plugins.

I have all the dependencies installed, and have done this:
git clone https://github.com/rabbitmq/rabbitmq-server.git rabbitmq
cd rabbitmq
git checkout v3.9.x
make

this does something, but errors for a reason that I can't decipher

<details><summary>`make` output</summary>

```
make[1]: Entering directory '/home/me/github/rabbitmq/deps/rabbit_common'
 DEP    jsx (3.1.0)
No beam files found.
Recompile: src/rebar
Recompile: src/rebar_abnfc_compiler
Recompile: src/rebar_app_utils
Recompile: src/rebar_appups
Recompile: src/rebar_asn1_compiler
Recompile: src/rebar_base_compiler
Recompile: src/rebar_cleaner
Recompile: src/rebar_config
Recompile: src/rebar_core
Recompile: src/rebar_cover_utils
Recompile: src/rebar_ct
src/rebar_ct.erl:291:30: Warning: crypto:rand_uniform/2 is deprecated; use rand:uniform/1 instead
%  291|     Random = integer_to_list(crypto:rand_uniform(0, 10000)),
%     |                              ^

Recompile: src/rebar_deps
Recompile: src/rebar_dia_compiler
Recompile: src/rebar_dialyzer
Recompile: src/rebar_edoc
Recompile: src/rebar_erlc_compiler
Recompile: src/rebar_erlydtl_compiler
Recompile: src/rebar_escripter
Recompile: src/rebar_eunit
src/rebar_eunit.erl:282:20: Warning: crypto:rand_uniform/2 is deprecated; use rand:uniform/1 instead
%  282|             Seed = crypto:rand_uniform(1, 65535),
%     |                    ^

Recompile: src/rebar_file_utils
Recompile: src/rebar_getopt
Recompile: src/rebar_lfe_compiler
Recompile: src/rebar_log
Recompile: src/rebar_metacmds
Recompile: src/rebar_mustache
Recompile: src/rebar_neotoma_compiler
Recompile: src/rebar_otp_app
Recompile: src/rebar_otp_appup
Recompile: src/rebar_port_compiler
Recompile: src/rebar_proto_compiler
Recompile: src/rebar_proto_gpb_compiler
Recompile: src/rebar_protobuffs_compiler
Recompile: src/rebar_qc
Recompile: src/rebar_rand_compat
Recompile: src/rebar_rel_utils
Recompile: src/rebar_reltool
Recompile: src/rebar_require_vsn
Recompile: src/rebar_shell
Recompile: src/rebar_subdirs
Recompile: src/rebar_templater
Recompile: src/rebar_upgrade
Recompile: src/rebar_utils
Recompile: src/rebar_xref
Recompile: src/rmemo
==> rebar (compile)
==> rebar (escriptize)
Congratulations! You now have a self-contained script called "rebar" in
your current working directory. Place this script anywhere in your path
and you can use rebar to build OTP-compliant apps.
/home/me/github/rabbitmq/deps/rabbit_common
make[2]: Entering directory '/home/me/github/rabbitmq/deps/jsx'
 DEPEND jsx.d
 ERLC   jsx.erl jsx_config.erl jsx_consult.erl jsx_decoder.erl jsx_encoder.erl jsx_parser.erl jsx_to_json.erl jsx_to_term.erl jsx_verify.erl
Old inliner: threshold=0 functions=[{doublequote,5},
                                    {singlequote,5},
                                    {maybe_replace,2},
                                    {format_number,1},
                                    {handle_event,3}]
 APP    jsx.app.src
make[2]: Leaving directory '/home/me/github/rabbitmq/deps/jsx'
Error: No Makefile to build dependency /home/me/github/rabbitmq/deps/recon.
make[1]: *** [erlang.mk:4512: deps] Error 2
make[1]: Leaving directory '/home/me/github/rabbitmq/deps/rabbit_common'
make: *** [erlang.mk:4512: deps] Error 2
```
</details>

I have very little hope of deciphering this error message or knowing what to do to resolve it. So there's no Makefile in the right place. Where should it have come from?

I'm quite frustrated - the Erlang changes were quite simple and took maybe 15 minutes but I can't actually build the plugin, despite having spent at this stage hours trying. This is just the most recent attempt.

Thanks for any help,
Chris P

**Comment by michaelklishin** [comment]
Recon is managed by rebar. `erlang.mk` should generate a Makefile for after it checks out the dependency.

I cannot know why that is not the case but you can delete `deps/recon` and then try to build the plugin:

```
cd deps/{plugin}
gmake
```

You must use a [compatible Erlang version](https://rabbitmq.com/which-erlang.html), for example, `24.3.4.13` for `v3.9.x`. It can be installed using `kerl` or `asdf`.

**Reply by chrispynz4** [reply]
Thanks,

I've been trying to build everything, not just a specific plugin and I don't have my plugin in this directory yet, only just what came from github v3.9.x branch.

I have RabbitMq itself 3.9.13 and Erlang 24.2.1 installed on this computer. They were installed from binaries, not from building the source. It's up and running and the management browser plugin is working, I can create exchanges, queues, etc so I think it's correctly installed. I'm eventually trying to build a plugin for RabbitMq 3.9 so I think I have the right things installed. Also python, elixir, etc from the dependencies page - they are also installed.

I removed deps/recon and tried again and get a similar problem:
Error: No Makefile to build dependency /home/me/github/rabbitmq/deps/credentials_obfuscation.

I notice you're using gmake rather than make. Is that what I'm doing wrong?
Alternately, can I just keep deleting from deps directory until it builds successfully?

Chris P

**Reply by chrispynz4** [reply]
If I run gmake instead of make there is no difference

**Reply by michaelklishin** [reply]
@chrispynz4 that likely means that you have previously tried building with an incompatible Erlang or Elixir version. erlang.mk does not re-fetch dependencies that failed to build.

``` shell
gmake distclean
```

in repository root should clear all non-committed dependencies IIRC. Or you can delete the manually like you did with `recon`.

This is one of the reasons why RabbitMQ has moved to Bazel for most things (except for release production, that still uses Make because all builds run in isolated environments).

**Comment by chrispynz4** [comment]
@michaelklishin 
I'm grateful for the help. I think we're getting closer.
Here's the latest:
me@cpubu2204:~/github/rabbitmq (git::v3.9.x)$ gmake distclean
 GEN    coverdata-clean
 GEN    clean-dist
 GEN    clean-ct-logs-archive
 GEN    clean-rabbitmq-deps.mk
 GEN    clean-source-dist
 GEN    distclean-tmp
 GEN    distclean-kerl
 GEN    distclean-ct
 GEN    distclean-plt
 GEN    distclean-edoc
 GEN    distclean-escript
 GEN    cover-report-clean
 GEN    distclean-packages
me@cpubu2204:~/github/rabbitmq (git::v3.9.x)$ gmake
gmake[1]: Entering directory '/home/me/github/rabbitmq/deps/rabbit_common'
 DEPEND rabbit_common.d
 ERLC   app_utils.erl code_version.erl credit_flow.erl delegate.erl delegate_sup.erl file_handle_cache.erl file_handle_cache_stats.erl gen_server2.erl mirrored_supervisor.erl mirrored_supervisor_locks.erl mnesia_sync.erl pmon.erl priority_queue.erl rabbit_amqp_connection.erl rabbit_amqqueue_common.erl rabbit_auth_backend_dummy.erl rabbit_auth_mechanism.erl rabbit_authn_backend.erl rabbit_authz_backend.erl rabbit_basic_common.erl rabbit_binary_generator.erl rabbit_binary_parser.erl rabbit_cert_info.erl rabbit_channel_common.erl rabbit_command_assembler.erl rabbit_control_misc.erl rabbit_core_metrics.erl rabbit_data_coercion.erl rabbit_date_time.erl rabbit_env.erl rabbit_error_logger_handler.erl rabbit_event.erl rabbit_exchange_type.erl rabbit_framing_amqp_0_8.erl rabbit_framing_amqp_0_9_1.erl rabbit_heartbeat.erl rabbit_http_util.erl rabbit_json.erl rabbit_log.erl rabbit_misc.erl rabbit_msg_store_index.erl rabbit_net.erl rabbit_nodes_common.erl rabbit_numerical.erl rabbit_password_hashing.erl rabbit_pbe.erl rabbit_peer_discovery_backend.erl rabbit_policy_validator.erl rabbit_queue_collector.erl rabbit_registry.erl rabbit_registry_class.erl rabbit_resource_monitor_misc.erl rabbit_runtime.erl rabbit_runtime_parameter.erl rabbit_semver.erl rabbit_semver_parser.erl rabbit_ssl_options.erl rabbit_types.erl rabbit_writer.erl supervisor2.erl vm_memory_monitor.erl worker_pool.erl worker_pool_sup.erl worker_pool_worker.erl
rabbit_cert_info.erl:10:14: can't find include lib "public_key/include/public_key.hrl"
rabbit_cert_info.erl:78:24: undefined macro 'id-ce-subjectAltName'
rabbit_cert_info.erl:133:15: undefined macro 'id-at-surname'
rabbit_cert_info.erl:12:2: function subject_alternative_names/1 undefined
rabbit_cert_info.erl:40:19: record 'OTPCertificate' undefined
rabbit_cert_info.erl:43:43: variable 'Issuer' is unbound
rabbit_cert_info.erl:50:19: record 'OTPCertificate' undefined
rabbit_cert_info.erl:53:43: variable 'Subject' is unbound
rabbit_cert_info.erl:61:19: record 'OTPCertificate' undefined
rabbit_cert_info.erl:64:42: variable 'Subject' is unbound
rabbit_cert_info.erl:67:37: record 'Extension' undefined
rabbit_cert_info.erl:69:19: record 'OTPCertificate' undefined
rabbit_cert_info.erl:72:23: variable 'Extensions' is unbound
rabbit_cert_info.erl:75:2: spec for undefined function subject_alternative_names/1
rabbit_cert_info.erl:88:19: record 'OTPCertificate' undefined
rabbit_cert_info.erl:91:72: variable 'Start' is unbound
rabbit_cert_info.erl:92:72: variable 'End' is unbound
rabbit_cert_info.erl:101:11: variable 'V' is unbound
rabbit_cert_info.erl:101:16: record 'AttributeTypeAndValue' undefined
rabbit_cert_info.erl:103:16: variable 'T' is unbound
rabbit_cert_info.erl:127:18: function format_rdn/1 undefined
rabbit_cert_info.erl:161:1: Warning: function escape_rdn_value/1 is unused
rabbit_cert_info.erl:164:1: Warning: function escape_rdn_value/2 is unused
gmake[2]: *** [erlang.mk:5456: ebin/rabbit_common.app] Error 1
gmake[1]: *** [erlang.mk:5198: app] Error 2
gmake[1]: Leaving directory '/home/me/github/rabbitmq/deps/rabbit_common'
gmake: *** [erlang.mk:4512: deps] Error 2

**Comment by michaelklishin** [comment]
I managed to compile `rabbitmq_message_timestamp` with Erlang 24.3.4.13 and Elixir 1.14.5-otp-24.

There were two changes necessary:

 * Bump `stdout_formatter` to include https://github.com/rabbitmq/stdout_formatter/commit/8c72e91fc8482e19872d83455cfdb688cebac071
 * Stop running `format --check-formatted` since that fails with code formatted for Elixir 1.14+ (https://github.com/rabbitmq/rabbitmq-server/commit/8760e7ccc79e9a29352904174752a5c0b15078c6)

**Reply by michaelklishin** [reply]
To get the latter all you have to do is to pull the `v3.9.x` branch.

**Reply by michaelklishin** [reply]
``` diff
diff --git a/deps/rabbitmq_cli/mix.exs b/deps/rabbitmq_cli/mix.exs
index c873f72762..ff74e30151 100644
--- a/deps/rabbitmq_cli/mix.exs
+++ b/deps/rabbitmq_cli/mix.exs
@@ -130,7 +130,7 @@ defmodule RabbitMQCtl.MixfileBase do
     elixir_deps = [
       {:json, "~> 1.4.1"},
       {:csv, "~> 2.4.0"},
-      {:stdout_formatter, "~> 0.2.3"},
+      {:stdout_formatter, git: "https://github.com/rabbitmq/stdout_formatter.git"},
       {:observer_cli, "~> 1.7.3"},
       {:amqp, "~> 2.1.0", only: :test},
       {:dialyxir, "~> 0.5", only: :test, runtime: false},
```

The following should make `stdout_formatter` compile on Erlang 24.3 with Elixir 1.14. I don't know if we can introduce a git dependency like that, so posting a patch that you can apply manually. And then nuke `deps/rabbitmq_cli/deps`.

**Comment by michaelklishin** [comment]
```
rabbit_cert_info.erl:10:14: can't find include lib "public_key/include/public_key.hrl"
```

Your Erlang installation is missing [TLS and related libraries](https://rabbitmq.com/ssl.html#erlang-otp-requirements). You are likely on Debian, where every part of the Erlang standard library is a separate package 🤦‍♂️

**Reply by chrispynz4** [reply]
Sort of. Ubuntu.

**Reply by chrispynz4** [reply]
I just tried to install of erlang-asn1, erlang-crypto, erlang-public-key, and erlang-ssl and in each case the attempt did nothing because they were already installed

**Reply by michaelklishin** [reply]
My best guess is that you use [a more recent version of Elixir than what `v3.9.x` can support, such as [in this thread](https://github.com/elixir-lang/elixir/issues/12681)).

Like I said, I used Elixir `1.14.5-otp-24` and Erlang `24.3.4.13` installed via `asdf`. It took
a couple of changes listed above (one was pushed to this repo on the `v3.9.x`) to get the broker
and CLI tools to build, and then building the plugin was trivial.

**Reply by chrispynz4** [reply]
This comment helped progress further.
https://github.com/elixir-lang/elixir/issues/12681#issuecomment-1699374362

"For future users, I had this error on Elixir 1.14.0 on Debian Bookworm stable. Updating ssl_verify_fun didn't resolve it but installing the **erlang-dev** apt package did."

I installed erlang-dev and the make went for quite a bit longer and gave a different error.
...a bunch of things were successful...
...
 ERLC   conf_parse.erl cuttlefish.erl cuttlefish_advanced.erl cuttlefish_bytesize.erl cuttlefish_conf.erl cuttlefish_datatypes.erl cuttlefish_duration.erl cuttlefish_duration_parse.erl cuttlefish_effective.erl cuttlefish_enum.erl cuttlefish_error.erl cuttlefish_escript.erl cuttlefish_flag.erl cuttlefish_generator.erl cuttlefish_mapping.erl cuttlefish_rebar_plugin.erl cuttlefish_schema.erl cuttlefish_translation.erl cuttlefish_unit.erl cuttlefish_util.erl cuttlefish_validator.erl cuttlefish_variable.erl cuttlefish_vmargs.erl
src/cuttlefish_unit.erl: **undefined parse transform 'eunit_autoexport'**
gmake[4]: *** [/home/me/github/rabbitmq/erlang.mk:5456: ebin/cuttlefish.app] Error 1
gmake[3]: *** [/home/me/github/rabbitmq/erlang.mk:5198: app] Error 2
gmake[3]: Leaving directory '/home/me/github/rabbitmq/deps/cuttlefish'
gmake[2]: *** [../../../../erlang.mk:4512: deps] Error 2
gmake[2]: Leaving directory '/home/me/github/rabbitmq/deps/rabbit/apps/rabbitmq_prelaunch'
gmake[1]: *** [../../erlang.mk:4465: apps] Error 2
gmake[1]: Leaving directory '/home/me/github/rabbitmq/deps/rabbit'
gmake: *** [erlang.mk:4512: deps] Error 2

**Comment by michaelklishin** [comment]
`v3.9.x` is completely out of support, so some observations and suggestions above is as much time as I'm willing to dedicate to it.

Moving to `v3.11.x` means that you will be running a version that is still updated and is reasonably up-to-date in terms of modern tools and dependencies.

**Reply by chrispynz4** [reply]
If only it was that easy :-) . I'd love to upgrade but it's not my choice and I'm not confident that my superiors would agree to it in a timely manner.

**Reply by chrispynz4** [reply]
Thanks @michaelklishin I understand that the help was voluntary. I'm not quite there but I'm further along and maybe can figure some things out
